search/views.py

```diff
@@ -318,12 +123,154 @@ def search(request):
+        # This means that the clustering computation will be triggered asynchronously.
+        # Moreover, the applied clustering filter will have no effect.
+        # Somehow, we should inform the user that the clustering results were not available yet, and that
+        # he should try again later to use a clustering facet.
+        return []
+
+    return sounds_from_requested_cluster
+
+
+def clustering_facet(request):
+    """Triggers the computation of the clustering, returns the state of processing or the clustering facet.
+    """
+    # pass the url query params for later sending it to the clustering engine
+    url_query_params_string = request.META['QUERY_STRING']
+    # remove existing cluster facet filter from the params since the returned cluster facets will include 
+    # their correspondinng cluster_id query parameter (done in the template)
+    url_query_params_string = re.sub(r"(&cluster_id=[0-9]*)", "", url_query_params_string)
+
+    result = cluster_sound_results(request, features=DEFAULT_FEATURES)
+
+    # check if computation is finished. If not, send computation state.
+    if result['finished']:
+        if result['result'] is not None:
+            results = result['result']
+            num_clusters = len(results)
+        else:
+             return JsonResponse({'status': 'failed'}, safe=False)
+    elif result['error']:
+        return JsonResponse({'status': 'failed'}, safe=False)
+    else:
+        return JsonResponse({'status': 'pending'}, safe=False)
+
+    # check if facet filters are present in the search query
+    # if yes, filter sounds from clusters
+    query_params, _, _ = search_prepare_parameters(request)
+    if query_params['filter_query']:
+        sound_ids_filtered = get_sound_ids_from_solr_query(query_params)
+        results = [[sound_id for sound_id in cluster if int(sound_id) in sound_ids_filtered] 
+                   for cluster in results]
+
+    num_sounds_per_cluster = [len(cluster) for cluster in results]
+    classes = {sound_id: cluster_id for cluster_id, cluster in enumerate(results) for sound_id in cluster}
+
+    # label clusters using most occuring tags
+    sound_instances = sounds.models.Sound.objects.bulk_query_id(map(int, classes.keys()))
+    sound_tags = {sound.id: sound.tag_array for sound in sound_instances}
+    cluster_tags = defaultdict(list)
+
+    # extract tags for each clusters and do not use query terms for labeling clusters
+    query_terms = {t.lower() for t in request.GET.get('q', '').split(' ')}
+    for sound_id, tags in sound_tags.iteritems():
+        cluster_tags[classes[str(sound_id)]] += [t.lower() for t in tags if t.lower() not in query_terms]
+
+    # count 3 most occuring tags
+    # we iterate with range(len(results)) to ensure that we get the right order when iterating through the dict 
+    cluster_most_occuring_tags = [' '.join(zip(*Counter(cluster_tags[cluster_id]).most_common(3))[0]) 
```

Comment:
can you split this into a few different lines, it's quite difficult to follow exactly what it's doing in a single line.